exporters/prometheus/exporter.go

```diff
@@ -0,0 +1,185 @@
+type metricData struct {
+	name             string
+	description      *prometheus.Desc
+	attributeValues  []string
+	valueType        prometheus.ValueType
+	value            float64
+	histogramCount   uint64
+	histogramSum     float64
+	histogramBuckets map[float64]uint64
+}
+
+func getMetricData(metrics metricdata.ResourceMetrics) []*metricData {
+	allMetrics := make([]*metricData, 0)
+	for _, scopeMetrics := range metrics.ScopeMetrics {
+		for _, m := range scopeMetrics.Metrics {
+			switch v := m.Data.(type) {
+			case metricdata.Histogram:
+				allMetrics = append(allMetrics, getHistogramMetricData(v, m)...)
+			case metricdata.Sum[int64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Sum[float64]:
+				allMetrics = append(allMetrics, getSumMetricData(v, m)...)
+			case metricdata.Gauge[int64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			case metricdata.Gauge[float64]:
+				allMetrics = append(allMetrics, getGaugeMetricData(v, m)...)
+			}
+		}
+	}
+
+	return allMetrics
+}
+
+func getHistogramMetricData(histogram metricdata.Histogram, m metricdata.Metrics) []*metricData {
+	dataPoints := make([]*metricData, 0)
+	for _, dp := range histogram.DataPoints {
+		keys, values := getAttrs(dp.Attributes)
+		desc := prometheus.NewDesc(m.Name, m.Description, keys, nil)
+		buckets := make(map[float64]uint64)
+		for i, bound := range dp.Bounds {
+			buckets[bound] = dp.BucketCounts[i]
+		}
+		md := &metricData{
+			name:             m.Name,
+			description:      desc,
+			attributeValues:  values,
+			valueType:        prometheus.UntypedValue,
+			histogramCount:   dp.Count,
+			histogramSum:     dp.Sum,
+			histogramBuckets: buckets,
+		}
+		dataPoints = append(dataPoints, md)
+	}
+	return dataPoints
+}
+
+func getSumMetricData[N int64 | float64](sum metricdata.Sum[N], m metricdata.Metrics) []*metricData {
+	dataPoints := make([]*metricData, 0)
+	for _, dp := range sum.DataPoints {
+		keys, values := getAttrs(dp.Attributes)
+		desc := prometheus.NewDesc(m.Name, m.Description, keys, nil)
+		md := &metricData{
+			name:            m.Name,
+			description:     desc,
+			attributeValues: values,
+			valueType:       prometheus.CounterValue,
+			value:           float64(dp.Value),
+		}
+		dataPoints = append(dataPoints, md)
+	}
+	return dataPoints
+}
+
+func getGaugeMetricData[N int64 | float64](gauge metricdata.Gauge[N], m metricdata.Metrics) []*metricData {
+	dataPoints := make([]*metricData, 0)
+	for _, dp := range gauge.DataPoints {
+		keys, values := getAttrs(dp.Attributes)
+		desc := prometheus.NewDesc(m.Name, m.Description, keys, nil)
+		md := &metricData{
+			name:            m.Name,
+			description:     desc,
+			attributeValues: values,
+			valueType:       prometheus.GaugeValue,
+			value:           float64(dp.Value),
+		}
+		dataPoints = append(dataPoints, md)
+	}
+	return dataPoints
+}
+
+func getAttrs(attrs attribute.Set) ([]string, []string) {
```

Comment:
We need to sanitize the keys to replace unsupported characters with `_`.

[Spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#metric-attributes)

Reply:
Is the regex for label keys in the spec right? It says in your link:

> Prometheus metric label keys are required to match the following regex: `[a-zA-Z_:]([a-zA-Z0-9_:])*`

But the [Prometheus docs](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) say

> Label names may contain ASCII letters, numbers, as well as underscores. They must match the regex `[a-zA-Z_][a-zA-Z0-9_]*`. Label names beginning with `__` are reserved for internal use.

so `:` isn't valid in Prometheus, and the otel spec doesn't mention anything about the reserved double-underscore prefix

Reply:
Ah, sorry.  Metric names can have `:`, but label names can't.  I was following the OpenMetrics spec for those: https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#abnf.  I'll update the spec.

Reply:
Opened https://github.com/open-telemetry/opentelemetry-specification/pull/2788 to update the spec.

Reply:
No problem, thanks for updating